src/indexClinicalData/index.ts

```diff
@@ -0,0 +1,84 @@
+import { esDonorId } from "./utils";
+import transformToEsDonor from "./transformToEsDonor";
+import programDonorStream from "./programDonorStream";
+import { toEsBulkIndexActions } from "elasticsearch";
+import { STREAM_CHUNK_SIZE } from "config";
+import { Client } from "@elastic/elasticsearch";
+import logger from "logger";
+import { EsDonorDocument, EsHit } from "./types";
+import esb from "elastic-builder";
+
+export const queryDocumentsByDonorIds = async (
+  donorIds: Array<string>,
+  client: Client,
+  indexName: string
+) => {
+  const esQuery = esb
+    .requestBodySearch()
+    .size(donorIds.length)
+    // appending .keyword is required when using termquery/termsquery on a field that's been declared as type keyword in the mapping!
+    .query(esb.termsQuery("donorId.keyword", donorIds));
+
+  const esHits: Array<EsHit> = await client
+    .search({
+      index: indexName,
+      body: esQuery,
+    })
+    .then((res) => res.body.hits.hits)
+    .catch((err) => {
+      logger.error("error in grabbing donors by id from Elasticsearch: ", err);
+      return [];
+    });
+
+  return esHits;
+};
+
+export default async (
+  programShortName: string,
+  targetIndexName: string,
+  esClient: Client
+) => {
+  const donorStream = programDonorStream(programShortName, {
+    chunkSize: STREAM_CHUNK_SIZE,
+  });
+  let chunksCount = 0;
+  for await (const chunk of donorStream) {
+    const timer = `streaming ${
+      chunk.length
+    } donor(s) from chunk #${chunksCount++} of program ${programShortName}`;
+    logger.profile(timer);
+
+    const esHits = await queryDocumentsByDonorIds(
+      chunk.map(esDonorId),
+      esClient,
+      targetIndexName
+    );
+
+    const donorIdDocumentPairs = esHits.map(
+      (hit) => [hit._source.donorId, hit] as [string, EsHit]
+    );
+
+    const preExistingDonorHits = Object.fromEntries(donorIdDocumentPairs);
+
+    const esDocuments = await Promise.all(
+      chunk.map((donor) => {
+        const donorId = esDonorId(donor);
+        if (preExistingDonorHits.hasOwnProperty(donorId)) {
+          return transformToEsDonor(
+            donor,
+            preExistingDonorHits[donorId]._source
+          );
+        } else return transformToEsDonor(donor);
+      })
+    );
```

Comment:
this doesn't break, but I don't think there's anything async here right? if so we can remove the `await Promise.all` :)

Reply:
doesn't `Promise.all` return one big mega-promise that needs to be resolved? Typescript is complaining when I remove the `await` .

Reply:
oh I see, `transformToEsDonors` is async for some reason... but doesn't look like there's anything async about it :/

Reply:
I noticed that too! I didn't know why it was async but I assumed it was like that pre-emptively in case the logic for transforming the RDPC needed to fetch something in the future?

Reply:
ah yeah that'd make sense... we're going down a pretty different route now so it wouldn't apply any more :P

Reply:
ok cool! thanks for pointing that out then, I'll convert it to a regular function!